versioned/persist/store/src/main/java/org/projectnessie/versioned/persist/store/PersistVersionStore.java

```diff
@@ -468,11 +468,30 @@ public PaginationIterator<KeyEntry> getKeys(Ref ref, String pagingToken)
 
     return new FilteringPaginationIterator<KeyListEntry, KeyEntry>(
         source.iterator(),
-        entry ->
-            KeyEntry.of(
-                DefaultStoreWorker.contentTypeForPayload(entry.getPayload()),
-                entry.getKey(),
-                entry.getContentId().getId())) {
+        entry -> {
+          if (withContent) {
+            try {
+              ContentAndState cs =
+                  databaseAdapter
+                      .values(
+                          hash, Collections.singleton(entry.getKey()), KeyFilterPredicate.ALLOW_ALL)
+                      .get(entry.getKey());
+              if (cs != null) {
+                Content content = mapContentAndState(cs);
+                return KeyEntry.of(
+                    DefaultStoreWorker.contentTypeForPayload(entry.getPayload()),
+                    entry.getKey(),
+                    content);
+              }
+            } catch (ReferenceNotFoundException e) {
+              throw new RuntimeException(e);
```

Comment:
nit: `IllegalStateException` is more concrete,  IMHO, since we expect the reference to be valid at this point.